pkg/platform/runtime/artifactvalidator/validator.go

```diff
@@ -0,0 +1,152 @@
+package artifactvalidator
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"strings"
+
+	"github.com/ActiveState/cli/internal/download"
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/fileutils"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"go.mozilla.org/pkcs7"
+)
+
+type signature struct {
+	Sig  string `json: "sig"`
+	Cert string `json: "cert"`
+}
+
+type attestation struct {
+	Payload    string      `json: "payload"`
+	Signatures []signature `json: "signatures"`
+}
+
+func ValidateAttestation(attestationFile string) error {
+	data, err := fileutils.ReadFile(attestationFile)
+	if err != nil {
+		return errs.Wrap(err, "Could not read attestation: "+attestationFile)
+	}
+
+	att := attestation{}
+	err = json.Unmarshal(data, &att)
+	if err != nil {
+		return errs.Wrap(err, "Could not unmarshal attestation")
+	}
+
+	if len(att.Signatures) == 0 {
+		return locale.NewError("validate_attestation_fail_no_signatures", "No signatures")
+	}
+
+	// Verify signing certificate.
+	pemBlock, _ := pem.Decode([]byte(att.Signatures[0].Cert))
+	if pemBlock == nil {
+		return errs.Wrap(err, "Unable to decode attestation certificate")
+	}
+
+	cert, err := x509.ParseCertificate(pemBlock.Bytes)
+	if err != nil {
+		return errs.Wrap(err, "Unable to parse attestation certificate")
+	}
+
+	intermediates := x509.NewCertPool()
+	addIntermediatesToPool(cert, intermediates)
+
+	opts := x509.VerifyOptions{
+		Roots:         nil, // use system root CAs
+		Intermediates: intermediates,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageCodeSigning},
+	}
+	_, err = cert.Verify(opts)
+	if err != nil {
+		return errs.Wrap(err, "Unable to validate certificate")
+	}
+
+	// Verify signature.
+	payload := make([]byte, len(att.Payload))
+	n, err := base64.StdEncoding.Decode(payload, []byte(att.Payload))
+	if err != nil {
+		return errs.Wrap(err, "Unable to decode attestation payload")
+	}
+	payload = payload[:n]
+	hash := sha256.New()
+	hash.Write(payload)
+	digest := hash.Sum(nil)
+
+	signature := make([]byte, len(att.Signatures[0].Sig))
+	n, err = base64.StdEncoding.Decode(signature, []byte(att.Signatures[0].Sig))
+	if err != nil {
+		return errs.Wrap(err, "Unable to decode attestation signature")
+	}
+	signature = signature[:n]
+
+	publicKey, ok := cert.PublicKey.(*rsa.PublicKey)
+	if !ok {
+		return locale.NewError("validate_attestation_fail_public_key", "Certificate's public key is not an expected RSA pubkey")
+	}
+	err = rsa.VerifyPSS(publicKey, crypto.SHA256, digest, signature, &rsa.PSSOptions{Hash: crypto.SHA256})
+	if err != nil {
+		return errs.Wrap(err, "Unable to validate signature")
+	}
+
+	// TODO: read payload artifact SHAs and validate them against downloaded artifact SHAs.
```

Comment:
As mentioned previously, there will be at least one follow up story for this entire task to hook everything up. Once we have attestations coming in from the platform inside artifacts, we can decide how to handle everything. Pete wants this preliminary code in sooner rather than later.

Reply:
Nathan has had a strong resistance to committing TODO comments. I'm not certain what he'd say here, but my guess is that he'd want a placeholder story filed that is to be fleshed out more later, and then link to the placeholder instead of having the TODO.

Reply:
Okay. I'm just going to remove the TODO and defer to Pete on this one with regard to the ticket to be filed.